### PR TITLE
Change CyFunction defaults to a PyObject

### PR DESCRIPTION
**Pull request by da-woods**
... instead of a struct.

A variety of advantages are listed in #4374 (alongside the
disadvantage that notably more code is generated for a cdef class
than a struct). (In principle there might be some caching possible
here to share the classes between similar functions)

(Draft because I'm mainly seeing what it ends up breaking, if anything)

### CODE REVIEW

**Comment by da-woods** [comment]
This seems to work without huge changes, and I think would have benefits for future maintainability (easier to make defaults pickleable and modifyable by introspection I think).

My main concern is that it generates a reasonable chunk of extra code for pretty much every function with defaults. I think the easiest solution would be to special-case the common case where all the defaults are PyObjects, and just store the defaults as a tuple there. I think that would cut down the extra code quite a bit without restricting usefulness.

I'll leave this as a draft for now because I don't think this change is immediately needed

**Comment by scoder** [comment]
It sounds like deduplicating the classes would be more generally useful and less special casing than adding tuples as an implementation (although that might still be an additional option later). We already do something similar for ctuples and the automatically generated C function wrappers.

Since the classes are purely meant for storage, we could even deduplicate a bit by type size, e.g. use void* for all pointer types and cast to the right type on access. That should reduce the number of extension types considerably. It might (just guessing) still be useful to distinguish between C pointers and objects, when it comes to pickling.

**Comment by da-woods** [comment]
> It sounds like deduplicating the classes would be more generally useful and less special casing than adding tuples as an implementation 

Yes probably. The main thing we'd lose is that the attribute names would match the argument names. But I don't think that's too important.

> we could even deduplicate a bit by type size, e.g. use void* for all pointer types and cast to the right type on access

I'm less sure about that. The don't store too many pointers (probably just `char*`) since the arguments generally must be coercable from PyObjects, so may be of limited benefit. And it probably breaks the auto-pickling and auto-construction of c++ classes that it was going to use

**Comment by da-woods** [comment]
I've added some caching to try to reduce the number of types created.

I had a look at ExprNodes.py to see how much difference it made. It turns out that basically every default value is a literal like `None` so there's actually only one defaults struct created in the whole file (both before and after this change), and therefore the caching makes no difference to ExprNodes.py. That may not be a completely typical use-case but it's an interesting observation.

**Comment by da-woods** [comment]
I'll mark this as "ready" (the remaining test failure is just the occasional division by 0 profiling time error). With that said, it's a fairly low priority improvement since it fixes fairly little by itself

File: Cython/Compiler/Symtab.py [hunk 1]
```diff
@@ -1362,6 +1364,31 @@ def declare_tuple_type(self, pos, components):
...
+            scope.is_defuults_class_scope = True
+
+            # zero pad the argument number so they can be sorted
+            num_zeros = math.floor(math.log10(len(components)))
+            format_str = "arg{0:0%dd}" % num_zeros
+            for n, type_ in enumerate(components):
+                scope.declare_var(EncodedString(format_str.format(n)), type_, None,
+                                  is_cdef = True)
+        return entry
+
```
Comment (scoder) [inline Cython/Compiler/Symtab.py:1391]: Let's avoid the lengthy indented block when we already know we're done.
```suggestion
    def declare_defaults_c_class(self, pos, components):
        # returns an entry (for the c-class)
        components = tuple(components)
        try:
            return self._cached_defaults_c_class_entries[components]
        except KeyError:
            pass

        cname = self.next_id(Naming.defaults_struct_prefix)
        cname = EncodedString(cname)
        entry = self._cached_defaults_c_class_entries[components] = self.declare_c_class(
            cname, pos, defining=True, implementing=True,
            objstruct_cname=cname)
        self.check_c_class(entry)
        entry.type.is_final_type = True
        scope = entry.type.scope
        scope.is_internal = True
        scope.is_defuults_class_scope = True

        # zero pad the argument number so they can be sorted
        num_zeros = math.floor(math.log10(len(components)))
        format_str = "arg{0:0%dd}" % num_zeros
        for n, type_ in enumerate(components):
            scope.declare_var(EncodedString(format_str.format(n)), type_, None, is_cdef = True)
        return entry

```

**Review by scoder — APPROVED** [review]

**Comment by da-woods** [comment]
I think I'm going to have another look at the code-size consequences of this before merging since I've become a little more worried about code size since writing this.

I think my main motivation was making the defaults pickleable (which was something I'd like in the future but isn't useful yet).

**Comment by da-woods** [comment]
I've had a bit of a look at the code-size - it does make some cases noticeably bigger (e.g. `def f(arg=[]): pass` alone ends up ~5% bigger).

With that said - I'm not too worried because:
* the deduplication of types
* theses are actually really rare - most defaults are just literals. I turned binding on and recompiled Cython itself and didn't spot any  significant code size change.
* I think the ability to pickle these will be nice in future.

**Pull request merged**